### PR TITLE
monsters: Monsters on patrol

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
Features "monsters: Monsters on patrol"

#### Purpose of change
Monsters should be able to patrol an area: move in a pre-defined route while diverging as needed to attack targets.  They can't currently, so add that functionality.

#### Describe the solution
Added a new to "spawn_data" called "patrol", which is a list of relative map square points.  The patrol points are supposed to be relative to the monster's OMT's (0, 0) position, but it's currently against the monster's pos().

When a monster with a patrol route is spawned, the patrol points are converted to a vector of absolute mapsquares and stored in the `monster` class.  The vector is saved and loaded as part of the `monster` class.

In `monster::plan()`, if the monster has no other activity, it sets its destination to the next patrol point or advances to the next patrol point if it within 2 mapsquares a patrol point.  The monster can't wander, so if it has nothing else to do, it paths to destination.

#### Describe alternatives you've considered
The first draft use the new co-ordinate types, but I ran into some issues.  Now that I have this working, ~~I may go back and use the new co-ordinate types.~~ The new co-ordinates are super confusing when shifting between the new types and the legacy types.

I want monsters on patrol to move on the overmap and enter the reality bubble, but that's a big task.

#### Testing
I created a map extra of 2-3 drones in a 72x72 mapsquare patrol, and observed the drones moving through the box.  The patrol pattern intercepted several trees, and the drones pathed around them.

I created a map extra of 2-3 drones in a 10x10 mapsquare patrol, and observed the drones moving through the box.  I also took an animated gif (see below).

I created a map extra of 2-3 zombies in a 10x10 mapsquare patrol, and observed the zombies moving through the box pattern.  I then spawned a motionless deer about 10 mapsquares off the patrol route.  The zombies moved to engage, killed the motionless deer, and then returned to the patrol.  I then edited the map to include a glass wall across their path, and the zombies managed to (eventually) path around it and continue the patrol.

#### Additional context
This PR includes #48136 but doesn't depend on it aside from easing debugging.

A pair of manhacks on a 10x10 box patrol
CDDA-DronesOnPatrol

##### TODO
- ~~Rewrite using proper co-ordinate types.~~ Nope
- [x] Relative co-ordinates are against the OMT (0, 0) position, not the monster's position.

### CODE REVIEW

File: doc/MAPGEN.md [hunk 1]
```diff
@@ -572,6 +573,31 @@ Example:
...
+| ---         | ---
+| ammo        | A list of objects, each of which has an `"ammo_id"` field and a `"qty"` list of two integers. The monster will spawn with items of "ammo_id", with at least the first number in the "qty" and no more than the second.
+| patrol      | A list of objects, each of which has an `"x"` field and a `"y"` field. Either value can be a range or a single number. The x,y co-ordinates define a patrol point as an relative mapsquare point offset from the (0, 0) local mapsquare of the overmap terrain tile that the monster spawns in. Patrol points are converted to absolute mapqaure tripoints inside the monster generator.
+
+Monsters with a patrol point list will move to each patrol point, in order, whenever they have no more pressing action to take on their turn. Upon reaching the last point in the patrol point list, the monster will continue on to the first point in the list.
+
+Example:
+```json
+"place_monster": [
+    { "monster": "mon_zombie", "x": 12, "y": 12, "spawn_data": { patrol: [ { "x": 12, "y": 12 } ] } }
```
Comment (BrettDong) [inline doc/MAPGEN.md:588]: Should `patrol` be quoted as a string?

Comment (mlangsdorf) [reply doc/MAPGEN.md:588]: See, this is why I put it up as a draft for review!

**Comment by actual-nh** [comment]
Is there currently any functionality for paths to be cached? This may increase the utility of such.

**Comment by mlangsdorf** [comment]
No path caching; paths can dynamically change because things can get in the way and it would be too much effort for too little pay-off.

**Comment by actual-nh** [comment]
BTW, do you think `spawn_data` could help with #46780?

**Comment by mlangsdorf** [comment]
as far as I can tell, `spawn_data` is for making sure that monsters have ammo, so yes.

**Pull request merged**